runtime/bundles/org.eclipse.e4.core.di/src/org/eclipse/e4/core/internal/di/AnnotationLookup.java

```diff
@@ -0,0 +1,185 @@
+/*******************************************************************************
+ * Copyright (c) 2023, 2023 Hannes Wellmann and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Hannes Wellmann - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.e4.core.internal.di;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.eclipse.e4.core.di.IInjector;
+import org.eclipse.e4.core.di.suppliers.IObjectDescriptor;
+import org.eclipse.e4.core.di.suppliers.PrimaryObjectSupplier;
+
+/**
+ * A utility class to ease the look-up of jakarta/javax.inject and
+ * jakarta/javax.annotation annotations and types as mutual replacements, while
+ * being able to handle the absence of javax-classes in the runtime.
+ *
+ * If support for javax-annotations is removed, this class can be simplified to
+ * only handle jakarta-annotations, then all method can be inlined and this
+ * class eventually deleted, together with the entire test-project
+ * org.eclipse.e4.core.tests.
+ */
+public class AnnotationLookup {
+	private AnnotationLookup() {
+	}
+
+	public static record AnnotationProxy(List<Class<? extends Annotation>> classes) {
+		public AnnotationProxy {
+			classes = List.copyOf(classes);
+		}
+
+		public boolean isPresent(AnnotatedElement element) {
+			for (Class<? extends Annotation> annotationClass : classes) {
+				if (element.isAnnotationPresent(annotationClass)) {
+					return true;
+				}
+			}
+			return false;
+		}
+	}
+
+	static final AnnotationProxy INJECT = createProxyForClasses(jakarta.inject.Inject.class,
+			() -> javax.inject.Inject.class);
+	static final AnnotationProxy SINGLETON = createProxyForClasses(jakarta.inject.Singleton.class,
+			() -> javax.inject.Singleton.class);
+	static final AnnotationProxy QUALIFIER = createProxyForClasses(jakarta.inject.Qualifier.class,
+			() -> javax.inject.Qualifier.class);
+
+	static final AnnotationProxy PRE_DESTROY = createProxyForClasses(jakarta.annotation.PreDestroy.class,
+			() -> javax.annotation.PreDestroy.class);
+	public static final AnnotationProxy POST_CONSTRUCT = createProxyForClasses(jakarta.annotation.PostConstruct.class,
+			() -> javax.annotation.PostConstruct.class);
```

Comment:
doesn't javax fail at runtime when javax is not present due to optional?

Reply:
That's right. But that is why this class is wrapped into a supplier, which is called in a try-catch block that catches NoClassDefFoundErrors.

The e4.core.tests explicitly got a case added to ensure that the `javax.inject/annotation` classes are not available in those tests-runtime so that a failure due to absense of those classes would show up in any of those tests.
And to ensure javax still works https://github.com/eclipse-platform/eclipse.platform/pull/791 was added.